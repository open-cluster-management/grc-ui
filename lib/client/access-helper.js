/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019, 2020. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

const filterTableAction = (data,actions,userAccessHash,resourceType) => {
  let actionList = []
  if(resourceType && resourceType.name && Array.isArray(actions)) {
    let adminRules = [], typeRules = []
    const namespace = (data && data.metadata && data.metadata.namespace)
      ? data.metadata.namespace
      : ''
    switch (resourceType.name) {
    case 'HCMCompliance':
    case 'HCMPolicyPolicy':
      if (namespace) {
        adminRules = userAccessHash[`${namespace}/*/*`]
        typeRules = userAccessHash[`${namespace}/policy.open-cluster-management.io/policies`]
        actionList = buildActionByUser(adminRules, typeRules, actions, actionList)
      }
      break
    case 'HCMPolicyCluster':
      return actions // all kind of cluster actions are available
    case 'policy.pp.details.title':
      if (namespace) {
        adminRules = userAccessHash[`${namespace}/*/*`]
        typeRules = userAccessHash[`${namespace}/apps.open-cluster-management.io/placementrules`]
        actionList = buildActionByUser(adminRules, typeRules, actions, actionList)
      }
      break
    case 'policy.pb.details.title':
      if (namespace) {
        adminRules = userAccessHash[`${namespace}/*/*`]
        typeRules = userAccessHash[`${namespace}/policy.open-cluster-management.io/placementbindings`]
        actionList = buildActionByUser(adminRules, typeRules, actions, actionList)
      }
      break
    default:
      // do nothing
    }
    return actionList
  }
}

function buildActionByUser(adminRules, typeRules, actions, actionList) {
  if (Array.isArray(adminRules) && adminRules.length > 0) {
    if (adminRules.includes('*')) { // if admin and *
      return actions // all kind of actions are available
    }
    else { // build action list based admin permission
      actionList = buildActionList(actions, adminRules, actionList)
    }
  }
  if (Array.isArray(typeRules) && typeRules.length > 0) {
    if (typeRules.includes('*')) { // if * for api on target NS
      return actions // all kind of api actions are available
    }
    else { // build action list based user api permission on target NS
      actionList = buildActionList(actions, typeRules, actionList)
    }
    return actionList
  }
}

function buildActionList(actions, rulesSet, actionList) {
  const removeFlag = actions.includes('table.actions.remove')
  const editFlag = actions.includes('table.actions.edit')
  const disableFlag = actions.includes('table.actions.disable')
  const enforceFlag = actions.includes('table.actions.enforce')
  if (actions.includes('table.actions.policy.policies.sidepanel')) {
    actionList.push('table.actions.policy.policies.sidepanel')
  }
  if (rulesSet.includes('update') || rulesSet.includes('patch')) {
    if (editFlag) {
      actionList.push('table.actions.edit')
    }
    if (disableFlag) {
      actionList.push('table.actions.disable')
    }
    if (enforceFlag) {
      actionList.push('table.actions.enforce')
    }
  } else {
    if (editFlag) {
      actionList.push('disabled.table.actions.edit')
    }
    if (disableFlag) {
      actionList.push('disabled.table.actions.disable')
    }
    if (enforceFlag) {
      actionList.push('disabled.table.actions.enforce')
    }
  }
  if (rulesSet.includes('delete') || rulesSet.includes('deletecollection')) {
    if (removeFlag) {
      actionList.push('table.actions.remove')
    }
  } else {
    if (removeFlag) {
      actionList.push('disabled.table.actions.remove')
    }
  }
  return actionList
}

export default filterTableAction
