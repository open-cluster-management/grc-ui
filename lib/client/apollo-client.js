/* eslint-disable no-console */
/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019, 2020. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import { ApolloClient, ApolloLink, HttpLink, InMemoryCache } from 'apollo-boost'
import { withClientState } from 'apollo-link-state'
import lodash from 'lodash'
import gql from 'graphql-tag'
import config from '../shared/config'
import * as Query from './queries'

const queryNewSearchStr = 'New Search (Unsaved)'

const defaultOptions = {
  watchQuery: {
    fetchPolicy: 'network-only',
    errorPolicy: 'all',
  },
  query: {
    fetchPolicy: 'network-only',
    errorPolicy: 'all',
  },
  mutate: {
    errorPolicy: 'all'
  }
}

const typeDefs = gql`
  scalar JSON

  type Modal {
    open: Boolean
    type: String
    data: JSON
  }

  type SearchQueryTab {
    queryName: String
    description: String
    searchText: String
    updated: Boolean
    id: String
  }

  type SearchQueryTabs {
    unsavedCount: Int
    openedTabName: String
    openedTabId: String
    tabs: [SearchQueryTab]
  }

  type Query {
    modal: Modal
    searchQueryTabs: SearchQueryTabs
  }

  type Mutation {
    updateModal(__typename: String, open: Boolean, type: String, data: JSON): JSON
    updateQueryTabs(__typename: String, unsavedCount: Int, openedTabName: String, openedTabId: String, data: JSON, tabs: [JSON]): JSON
    updateSingleQueryTab(openedTabName: String, openedTabId: String, description: String, searchText: String, updateUnsavedOrExisting: Boolean): JSON
    removeSingleQueryTab(id: String): JSON
  }
`

const stateLink = withClientState({
  cache: new InMemoryCache(),
  defaults: {
    searchInput: {
      __typename: 'SearchInput',
      text: '',
    },
    modal: {
      __typename: 'modal',
      open: false,
      type:'',
      data: {
        __typename: 'ModalData',
        name: '',
        searchText:'',
        description:''
      }
    },
    searchQueryTabs: {
      __typename: 'SearchQueryTabs',
      tabs: [{
        __typename: 'QueryTab',
        queryName: queryNewSearchStr,
        searchText:'',
        description:'',
        updated: false,
        id: queryNewSearchStr
      }],
      unsavedCount: 1,
      openedTabName: queryNewSearchStr,
      openedTabId: queryNewSearchStr
    },
    relatedResources: {
      __typename: 'RelatedResources',
      visibleKinds: []
    }
  },
  typeDefs: typeDefs,
  resolvers: {
    Mutation: {
    },
  }
})

const getXsrfToken = () => {
  const token = document.getElementById('app-access') ? document.getElementById('app-access').value : ''
  return token.toString('ascii')
}

const client = new ApolloClient({
  connectToDevTools: true,
  cache: new InMemoryCache(),
  link: ApolloLink.from([stateLink, new HttpLink({
    uri: `${config.contextPath}/graphql`,
    credentials: 'same-origin',
    headers: {
      'XSRF-Token': getXsrfToken(),
      'context': {
        'locale': 'en'
      }
    },
  }) ]),
  defaultOptions,
})

const searchClient = new ApolloClient({
  connectDevTools: process.env.NODE_ENV === 'development',
  cache: new InMemoryCache(),
  link: ApolloLink.from([
    withClientState({
      cache: new InMemoryCache(),
    }), new HttpLink({
      uri: `${config.contextPath}/search/graphql`,
      credentials: 'same-origin',
      headers: {
        'XSRF-Token': getXsrfToken(),
      },
    })
  ]),
  defaultOptions,
})

class GrcApolloClient {

  getGrcClient(){
    return client
  }

  getSearchClient() {
    // TODO - use flag while ironing out the chart changes
    return config['feature_search-api'] ? searchClient : client
  }

  // general search using search-api
  search(q, variables = {}) {
    // TODO - use flag while ironing out the chart changes
    return config['feature_search-api'] ? searchClient.query({ query: q, variables }) : client.query({ query: q, variables })
  }

  // gets list
  get(resourceType, variables = {}) {
    const resourceList = lodash.get(Query, resourceType.list)
    if (resourceList) {
      return client.query({ query: resourceList, variables })
    }
    return Promise.resolve()
  }

  // gets one resource
  getResource(resourceType, variables = {}) {
    return client.query({ query: lodash.get(Query, resourceType.name), variables })
  }
  createPolicy(resources) {
    return client.mutate({ mutation: Query.createPolicy, variables: { resources } })
  }
  createCompliance(resources) {
    return client.mutate({ mutation: Query.createCompliance, variables: { resources } })
  }
  createApplication(resources){
    return client.mutate({ mutation: Query.createApplication, variables: { resources } })
  }
  saveUserQueries(resource){
    return client.mutate({ mutation: Query.createUserQueries, variables: { resource } })
  }
  createResources(resources){
    return client.mutate({ mutation: Query.createResources, variables: { resources } })
  }
  createAndUpdateResources(toCreate, toUpdate){
    return client.mutate({ mutation: Query.createAndUpdateResources, variables: { toCreate, toUpdate } })
  }

  updateResource(resourceType, namespace, name, body, selfLink, resourcePath) {
    return client.mutate({ mutation: Query.updateResource, variables: { resourceType, namespace, name, body, selfLink, resourcePath } })
  }
  updateResourceLabels(resourceType, namespace, name, body, selfLink, resourcePath) {
    return client.mutate({ mutation: Query.updateResourceLabels, variables: { resourceType, namespace, name, body, selfLink, resourcePath } })
  }
  deleteUserQueries(resource) {
    return client.mutate({ mutation: Query.removeQuery, variables: {resource}})
  }

  remove(input) {
    const removeKind = input.kind
    switch(removeKind) {
    case 'HCMPolicyPolicy':
    default:{
      const comRes = input.selected.filter((res) => res.selected === true)
      return client.mutate({ mutation: Query.deleteCompliance, variables: { selfLink: input.metadata.selfLink, resources: comRes } })
    }
    case 'HCMSecurityFindings': {
      if (input.name) {
        return client.mutate({ mutation: Query.deleteOccurrences, variables: { selfLink: input.name} })
      }
    }
    }
    return undefined
  }
}

export default new GrcApolloClient()
