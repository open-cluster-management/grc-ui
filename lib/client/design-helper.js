/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import jsYaml from 'js-yaml'
import YamlParser from '../../lib/client/yaml-helper'
import msgs from '../../nls/platform.properties'
import _ from 'lodash'

// itemKeys: members of item that get their own chart section
export const dumpAndParse = (item, itemKeys=[]) => {

  const tempItem = _.cloneDeep(item)

  const removeStatus = (obj) => {
    if (obj.raw) {
      delete obj.raw['status']
    }
  }
  removeStatus(tempItem)
  // use this to sort the keys generated by safeDump
  const sortKeys = (a,b) => {
    if (a==='name' && b!=='name') {
      return -1
    } else if (a!=='name' && b==='name') {
      return 1
    } else if (a==='namespace' && b!=='namespace') {
      return -1
    } else if (a!=='namespace' && b==='namespace') {
      return 1
    }
    return a.localeCompare(b)
  }

  // dump parsed data into yaml sections
  let yaml, row=0
  let parsed
  const yamls = []
  if (tempItem.raw) {
    yaml = jsYaml.safeDump(tempItem.raw, {sortKeys})
    // create objects from the dumped yaml to create with row #'s
    parsed = new YamlParser().parse(yaml)
    row += yaml.split('\n').length
    yamls.push(yaml)
  }

  // then dump the arrays within parsed
  itemKeys.forEach(key=>{
    const arr = tempItem[key]
    if (Array.isArray(arr)) {
      parsed[key] = []
      arr.forEach(member=>{
        if (member.raw) {
          yaml = jsYaml.safeDump(member.raw, {sortKeys})
          // create objects from the dumped yaml to create with row #'s
          const synced = new YamlParser().parse(yaml, row)
          synced.$r = row
          parsed[key].push(synced)
          row += yaml.split('\n').length
          yamls.push(yaml)
        }
      })
    }
  })
  return {yaml: yamls.join('---\n'), parsed}
}

export const saveLoad = (data) => {
  return jsYaml.safeLoadAll(data)
}

export const getResourceType = (item, locale, key) => {
  return key ? _.get(item, key) : item.resourceType
}

export const getClusterLink = (item) => {
  const { cluster: { consoleURL = '', metadata} } = item

  return consoleURL !== '' ?
    <a target='_blank' rel='noopener noreferrer' href={`${consoleURL}/console`}>{metadata.name}</a>
    : metadata.name
}

export function transformPVStatus(item, locale) {
  const translatedStatus = msgs.get(`table.cell.status.${item.status.toLowerCase()}`, locale)
  return translatedStatus ===  `table.cell.status.${item.status.toLowerCase()}`
    ? translatedStatus
    : item.status
}

export const parse = (yaml, validator, locale) => {
  let absLine=0
  const parsed = {}
  const yamls = yaml.split(/^---$/gm)
  const exceptions = []
  // check for syntax errors
  try {
    yamls.forEach((snip)=>{
      const obj = jsYaml.safeLoad(snip)
      const key = _.get(obj, 'kind', 'unknown')
      let values = parsed[key]
      if (!values) {
        values = parsed[key] = []
      }
      const post = snip.endsWith('\n')
      snip = snip.trim()
      const $synced = new YamlParser().parse(snip, absLine)
      $synced.$r = absLine
      $synced.$l = snip.split('\n').length
      values.push({$raw: obj, $yml: snip, $synced})
      absLine += $synced.$l
      if (post) {
        absLine++
      }
    })
  } catch (e) {
    const {mark={}, reason, message} = e
    const {line=0, column=0} = mark
    exceptions.push({
      row: line+absLine,
      column,
      text: _.capitalize(reason||message),
      type: 'error',
    })
  }

  // validate
  if (validator && exceptions.length===0) {
    validator(parsed, exceptions, locale)
  }

  return {parsed, exceptions}
}
