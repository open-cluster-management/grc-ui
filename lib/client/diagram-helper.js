/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import ReactDOMServer from 'react-dom/server'

/*
* UI helpers to help with data transformations
* */

export const getWrappedNodeLabel = (label, width, rows=3) => {
  // if too long, add elipse and split the rest
  if (label.length>width*rows) {
    if (rows===2) {
      label = label.substr(0, width)+ '..\n' + label.substr(-width)
    } else {
      label = splitLabel(label.substr(0, width*2), width, rows-1) + '..\n' +  label.substr(-width)
    }
  } else {
    label = splitLabel(label, width, rows)
  }
  return label
}

const splitLabel = (label, width, rows) => {
  let line=''
  const lines = []
  const parts = label.split(/([^A-Za-z0-9])+/)
  let remaining = label.length
  do {
    // add label part
    line += parts.shift()

    // add splitter
    if (parts.length) {
      line += parts.shift()
    }

    // if next label part puts it over width split it
    if (parts.length) {
      if (line.length+parts[0].length > width) {
        remaining -= line.length
        if (remaining>width) {
          if (rows===2) {
            // if pentulitmate row do a hard break
            const split = parts[0]
            const idx = width - line.length
            line += split.substr(0,idx)
            parts[0] = split.substr(idx)
          }
        }
        lines.push(line)
        line = ''
        rows-=1
      }
    } else {
      // nothing left, push last line
      lines.push(line)
    }
  } while (parts.length)

  // pull last line in if too short
  if (lines.length>1) {
    let lastLine = lines.pop()
    if (lastLine.length<=2) {
      lastLine = lines.pop() + lastLine
    }
    lines.push(lastLine)
  }
  return lines.join('\n')
}

//as scale decreases from max to min, return a counter zoomed value from min to max
export const counterZoom = (scale, scaleMin, scaleMax, valueMin, valueMax) => {
  if (scale>=scaleMax) {
    return valueMin
  } else if (scale<=scaleMin) {
    return valueMax
  }
  return valueMin + (1-((scale-scaleMin)/(scaleMax-scaleMin))) * (valueMax-valueMin)
}

export const getTooltip = (tooltips) => {
  return tooltips.map(({name, value})=>{
    return ReactDOMServer.renderToStaticMarkup(<div>
      <span className='label'>{name}:  </span>
      <span className='value'>{value}</span>
    </div>)
  }).join('')
}

export const getHashCode = (str) => {
  let hash = 0, i, chr
  for (i = 0; i < str.length; i++) {
    chr   = str.charCodeAt(i)
    hash  = ((hash << 5) - hash) + chr
    hash |= 0
  }
  return hash
}
