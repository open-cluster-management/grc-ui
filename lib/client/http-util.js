/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

const client = require('../shared/client')
const config = require('../shared/config')

const DEFAULT_OPTIONS = {
  credentials: 'same-origin',
  headers: {
    'X-Requested-With': 'XMLHttpRequest'
  }
}

module.exports.getHostUrl = function () {
  const port = window.location.port ? `:${window.location.port}` : ''
  return `${window.location.protocol}//${window.location.hostname}${port}`
}

module.exports.getContextRoot = function() {
  if (client) {
    // eslint-disable-next-line no-undef
    return CONSOLE_CONTEXT_URL
  }
  return config.contextPath
}

module.exports.fetch = function (url, successCb, errorCb, options) {

  options = options || DEFAULT_OPTIONS
  fetch(url, options)
    .then(checkStatus)
    .then(parseJSON)
    .then(json => successCb && successCb(json))
    .catch(ex => errorCb && errorCb({ error: ex }))
}

function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response
  } else {
    const error = new Error(response.statusText)
    error.response = response
    throw error
  }
}

function parseJSON(response) {
  // '' and 'OK' is not a valid json response, need to check
  return response.text().then(text => (text && text.trim() !== 'OK') ? JSON.parse(text) : {})
}
