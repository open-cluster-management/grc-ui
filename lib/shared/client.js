/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

/*
 * A boolean module that is true on the client, false on the server. This is for code that needs to
 * do different things on the server vs the client.
 */

module.exports = (typeof window !== 'undefined')
