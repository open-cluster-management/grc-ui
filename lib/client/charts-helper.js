/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import client from '../../lib/shared/client'

export const inflateKubeValue = value => {
  const match = value.match(/\D/g)
  if (match) {
    // if value has suffix
    const unit = match.join('')
    const val = value.match(/\d+/g).map(Number)[0]
    const BINARY_PREFIXES = ['Ki', 'Mi', 'Gi', 'Ti', 'Pi', 'Ei']
    const SI_PREFIXES = ['m', 'k', 'M', 'G', 'T', 'P', 'E' ]
    const num = unit && unit.length === 2 ? factorize(BINARY_PREFIXES, unit, 'binary') : factorize(SI_PREFIXES, unit, 'si')
    return val * num
  }
  return parseFloat(value)
}

function factorize(prefixes, unit, type) {
  let localFactorize = 1
  for (let index = 0; index < prefixes.length; index++) {
    if (unit === prefixes[index]) {
      const base = type === 'binary' ? 1024 : 1000
      const exponent = type === 'binary' ? (index + 1) : (unit === 'm' ? -1 : index)
      localFactorize = Math.pow(base, exponent)
    }
  }
  return localFactorize
}

export const deflateKubeValue = (size, locale, decimals, threshold, multiplier, units) => {
  size = size || 0
  if (decimals === null) {
    decimals = 2
  }
  threshold = threshold || 800 // Steps to next unit if exceeded
  multiplier = multiplier || 1024
  units = units || locale === 'fr' ? ['', 'Kio', 'Mio', 'Gio', 'Tio', 'Pio'] : ['', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB']

  let localFactorize = 1, unitIndex

  for (unitIndex = 0; unitIndex < units.length; unitIndex++) {
    if (unitIndex > 0) {
      localFactorize = Math.pow(multiplier, unitIndex)
    }

    if (size < multiplier * localFactorize && size < threshold * localFactorize) {
      break
    }
  }

  if (unitIndex >= units.length) {
    unitIndex = units.length - 1
  }

  let fileSize = size / localFactorize

  fileSize = client ? new Intl.NumberFormat(locale).format(fileSize.toFixed(decimals)) : fileSize.toFixed(decimals)

  // This removes unnecessary 0 or . chars at the end of the string/decimals
  if (fileSize.indexOf('.') > -1) {
    fileSize = fileSize.replace(/\.?0*$/, '')
  }

  return {size: parseFloat(fileSize), units: units[unitIndex]}
}

export const getPercentage = (value, total) => {
  return Math.floor(100 * value / total) || 0
}

export const getTotal = data => {
  return data.reduce((acc, item) => {
    return acc + item.value
  }, 0)
}
