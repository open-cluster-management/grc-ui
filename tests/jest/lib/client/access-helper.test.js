/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import { fliterTableAction } from '../../../../lib/client/access-helper'

const policyActions = [
  'table.actions.policy.policies.sidepanel',
  'table.actions.edit',
  'table.actions.disable',
  'table.actions.enforce',
  'table.actions.remove'
]

const resourceType = {
  'name':'HCMPolicyPolicy',
  'list':'HCMPolicyPolicyList'
}

// adminAccessDefault has all permission on default namespace
// adminAccessDefault only has view permission on calamari namespace
const adminAccessDefault = {
  'calamari/policy.open-cluster-management.io/policies': [
    'get',
    'list',
    'watch',
  ],
  'default/*/*': [
    '*'
  ],
}

describe('adminAccessDefault action list on different namespaces ', () => {
  it('should get all permission on default namespace', () => {
    expect(fliterTableAction(
      {
        metadata: {
          namespace: 'default'
        }
      },
      policyActions,
      adminAccessDefault,
      resourceType
    )).toMatchSnapshot()
  })

  it('should only get only has view permission on calamari namespace', () => {
    expect(fliterTableAction(
      {
        metadata: {
          namespace: 'calamari'
        }
      },
      policyActions,
      adminAccessDefault,
      resourceType
    )).toMatchSnapshot()
  })
})

// test1AccessDefault has all permission on calamari namespace
// test1AccessDefault only has edit permission on avengers namespace
// test1AccessDefault only has view permission on ironman namespace
const test1AccessDefault = {
  'calamari/policy.open-cluster-management.io/policies': [
    'get',
    'list',
    'watch',
    'update',
    'patch',
    'create',
    'delete',
    'deletecollection'
  ],
  'avengers/policy.open-cluster-management.io/policies': [
    'get',
    'list',
    'watch',
    'update',
    'patch',
  ],
  'ironman/policy.open-cluster-management.io/policies': [
    'get',
    'list',
    'watch',
  ],
}

describe('test1AccessDefault action list on different namespaces ', () => {
  it('should get all permission on calamari namespace', () => {
    expect(fliterTableAction(
      {
        metadata: {
          namespace: 'calamari'
        }
      },
      policyActions,
      test1AccessDefault,
      resourceType
    )).toMatchSnapshot()
  })
  it('should only get only has edit permission on avengers namespace', () => {
    expect(fliterTableAction(
      {
        metadata: {
          namespace: 'avengers'
        }
      },
      policyActions,
      test1AccessDefault,
      resourceType
    )).toMatchSnapshot()
  })
  it('should only get only has view permission on ironman namespace', () => {
    expect(fliterTableAction(
      {
        metadata: {
          namespace: 'ironman'
        }
      },
      policyActions,
      test1AccessDefault,
      resourceType
    )).toMatchSnapshot()
  })
})
