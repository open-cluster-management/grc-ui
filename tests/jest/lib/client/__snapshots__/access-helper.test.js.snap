// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adminAccessDefault action list on different namespaces  should get all permission on default namespace 1`] = `
Array [
  "table.actions.policy.policies.sidepanel",
  "table.actions.edit",
  "table.actions.disable",
  "table.actions.enforce",
  "table.actions.remove",
]
`;

exports[`adminAccessDefault action list on different namespaces  should only get only has view permission on calamari namespace 1`] = `
Array [
  "table.actions.policy.policies.sidepanel",
]
`;

exports[`test1AccessDefault action list on different namespaces  should get all permission on calamari namespace 1`] = `
Array [
  "table.actions.policy.policies.sidepanel",
  "table.actions.remove",
  "table.actions.edit",
  "table.actions.disable",
  "table.actions.enforce",
]
`;

exports[`test1AccessDefault action list on different namespaces  should only get only has edit permission on avengers namespace 1`] = `
Array [
  "table.actions.policy.policies.sidepanel",
  "table.actions.edit",
  "table.actions.disable",
  "table.actions.enforce",
]
`;

exports[`test1AccessDefault action list on different namespaces  should only get only has view permission on ironman namespace 1`] = `
Array [
  "table.actions.policy.policies.sidepanel",
]
`;
